pkg/kail/logger.go

```diff
@@ -0,0 +1,121 @@
+ * limitations under the License.
+ */
+
+package kail
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"time"
+
+	"github.com/boz/go-logutil"
+	"github.com/boz/kail"
+	"github.com/projectriff/riff/pkg/k8s"
+	"github.com/projectriff/system/pkg/apis/build"
+	buildv1alpha1 "github.com/projectriff/system/pkg/apis/build/v1alpha1"
+	"github.com/projectriff/system/pkg/apis/request"
+	requestv1alpha1 "github.com/projectriff/system/pkg/apis/request/v1alpha1"
+	"github.com/projectriff/system/pkg/apis/stream"
+	streamv1alpha1 "github.com/projectriff/system/pkg/apis/stream/v1alpha1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+)
+
+type Logger interface {
+	ApplicationLogs(ctx context.Context, application *buildv1alpha1.Application, since time.Duration, out io.Writer) error
+	FunctionLogs(ctx context.Context, function *buildv1alpha1.Function, since time.Duration, out io.Writer) error
+	HandlerLogs(ctx context.Context, handler *requestv1alpha1.Handler, since time.Duration, out io.Writer) error
+	ProcessorLogs(ctx context.Context, processor *streamv1alpha1.Processor, since time.Duration, out io.Writer) error
+}
+
+func NewDefault(k8s k8s.Client) Logger {
+	return &logger{
+		k8s: k8s,
+	}
+}
+
+type logger struct {
+	k8s k8s.Client
+}
+
+func (c *logger) ApplicationLogs(ctx context.Context, application *buildv1alpha1.Application, since time.Duration, out io.Writer) error {
+	selector, err := labels.Parse(fmt.Sprintf("%s=%s", build.ApplicationLabelKey, application.Name))
+	if err != nil {
+		panic(err)
+	}
+	containers := []string{}
+	return c.stream(ctx, application.Namespace, selector, containers, since, out)
+}
+
+func (c *logger) FunctionLogs(ctx context.Context, function *buildv1alpha1.Function, since time.Duration, out io.Writer) error {
+	selector, err := labels.Parse(fmt.Sprintf("%s=%s", build.FunctionLabelKey, function.Name))
+	if err != nil {
+		panic(err)
+	}
+	containers := []string{}
+	return c.stream(ctx, function.Namespace, selector, containers, since, out)
+}
+
+func (c *logger) HandlerLogs(ctx context.Context, handler *requestv1alpha1.Handler, since time.Duration, out io.Writer) error {
+	selector, err := labels.Parse(fmt.Sprintf("%s=%s", request.HandlerLabelKey, handler.Name))
+	if err != nil {
+		panic(err)
+	}
+	containers := []string{"user-container"}
+	return c.stream(ctx, handler.Namespace, selector, containers, since, out)
+}
+
+func (c *logger) ProcessorLogs(ctx context.Context, processor *streamv1alpha1.Processor, since time.Duration, out io.Writer) error {
+	selector, err := labels.Parse(fmt.Sprintf("%s=%s", stream.ProcessorLabelKey, processor.Name))
+	if err != nil {
+		panic(err)
+	}
+	containers := []string{"function", "processor"}
+	return c.stream(ctx, processor.Namespace, selector, containers, since, out)
+}
+
+func (c *logger) stream(ctx context.Context, namespace string, selector labels.Selector, containers []string, since time.Duration, out io.Writer) error {
```

Comment:
This is implemented using kail, but I thought kail was removed on the basis that it was faulty (duplicating or omitting logs). What's the rationale for risking kail again?

Reply:
kail does still miss some logs and duplicate others. I'm reintroducing it to restore the functionality. All kail usage is encapsulated within this package, so we can swap it out with something better in the future.